python-checks/src/main/java/org/sonar/python/checks/TfFunctionDependOnOutsideVariableCheck.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.AssignmentStatement;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.checks.utils.Expressions;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6911")
+public class TfFunctionDependOnOutsideVariableCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE = "Make sure this function does not depend on a global or free variable.";
+  private static final String MESSAGE_SECONDARY = "Variable used here.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, TfFunctionDependOnOutsideVariableCheck::checkFunction);
+  }
+
+  private static void checkFunction(SubscriptionContext context) {
+    FunctionDef functionDef = (FunctionDef) context.syntaxNode();
+
+    if (!TreeUtils.isFunctionWithGivenDecoratorFQN(functionDef, "tensorflow.function")) {
+      return;
+    }
+    NameCollector collector = new NameCollector();
+    functionDef.body().accept(collector);
+    Set<Symbol> allNames = new HashSet<>(collector.symbolToNames.keySet());
+    allNames.removeAll(functionDef.localVariables());
+    if (allNames.isEmpty()) {
+      return;
+    }
+    var issue = context.addIssue(functionDef.name(), MESSAGE);
+    allNames.forEach(symbol -> collector.symbolToNames.get(symbol).forEach(name -> issue.secondary(name, MESSAGE_SECONDARY)));
+  }
+
+  private static class NameCollector extends BaseTreeVisitor {
+    Map<Symbol, List<Name>> symbolToNames = new HashMap<>();
+
+    @Override
+    public void visitName(Name pyNameTree) {
```

Comment:
Would it make sense for this method to have access of the local variables of the enclosing function and to early return if the current symbol belongs to the function?
Then we'd only look at usages for symbols that were actually defined outside the function.